File: edoor/edoor/report/monthly_property_data_summary/monthly_property_data_summary.py
# Copyright (c) 2023, Tes Pheakdey and contributors
# For license information, please see license.txt
from frappe.utils import getdate, add_to_date
import frappe
from edoor.edoor.report.monthly_property_data_summary import report_summary_by_occupy
from edoor.edoor.report.utils import get_months

def execute(filters=None):
	if getdate(filters.start_date) > getdate(filters.end_date):
		frappe.throw("Start date cannot less than end date")

	min_max_day = get_min_max_day(filters)
	columns = get_report_columns(filters,min_max_day)
 
	report_data = {}
	if filters.column_group=="Occupy":
		report_data = report_summary_by_occupy.get_report(filters,columns)
	else:
		frappe.throw("This option is comming soon")
	message = ""
	if filters.row_group == "Room":
		message = "Note: Unassigned rooms in your reservation may affect the accuracy of the occupancy calculation."
	
	return columns, report_data["data"],message,report_data["report_chart"],report_data["report_summary"]

def get_report_columns(filters,min_max_day):
	columns = [
		{"fieldname": "row_group", "label":filters.row_group, "width": 200},
	]

	for n in range(min_max_day["min_day"], min_max_day["max_day"]+1):
		columns.append({
			"fieldname":"col_" + str(n),"label": "{}".format(n), "width":50, "align":"center", "has_total":True
	})
  
	columns.append({
		"fieldname":"occupancy","label": "Occ", "width":100, "align":"center","fieldtype":"Percent"
	})
 
	columns.append({
		"fieldname":"total","label": "Total", "width":100, "align":"center","has_total":True
	})
	
		
	return columns


def get_min_max_day(filters):
	sql = "select min(day(date)) as min_day, max(day(date)) as max_day from `tabDates` where date between %(start_date)s and %(end_date)s"
	data = frappe.db.sql(sql,filters,as_dict=1)
	return data [0]

File: edoor/edoor/report/monthly_property_data_summary/monthly_property_data_summary.js
// Copyright (c) 2024, Tes Pheakdey and contributors
// For license information, please see license.txt


frappe.query_reports["Monthly Property Data Summary"] = {
	"filters": [
		{
			fieldname: "property",
			label: "Property",
			fieldtype: "Link",
			options: "Business Branch",
			default: frappe.defaults.get_user_default("business_branch"),
			"reqd": 1,
			"on_change": function (query_report) {
				setLinkField()
			},
		},
		{
			"fieldname": "start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			default: new Date((new Date()).getFullYear(), (new Date()).getMonth(), 1),
			"reqd": 1,
			"on_change": function (query_report) { },
		},
		{
			"fieldname": "end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			default: new Date((new Date()).getFullYear(), (new Date()).getMonth() + 1, 0),
			"on_change": function (query_report) { },
			"reqd": 1
		},
		// {
		// 	"fieldname": "building",
		// 	"label": __("Building"),
		// 	"fieldtype": "Link",
		// 	"options": "Building",
		// 	"on_change": function (query_report) { },
		// },
		{
			"fieldname": "row_group",
			"label": __("Group By"),
			"fieldtype": "Select",
			"options": "Business Source\nBusiness Source Type\nReservation Type\nGuest Type\nNationality\nRoom\nRoom Type",
			"default": "Business Source",
			"on_change": function (query_report) { },
			hide_in_filter: 1,
		},
		{
			"fieldname": "column_group",
			"label": __("Column Group"),
			"fieldtype": "Select",
			"options": "Occupy\nRevenue\nADR",
			"default": "Occupy",
			"on_change": function (query_report) { },
			hide_in_filter: 1,
		},
		
		{
			"fieldname": "chart_type",
			"label": __("Chart Type"),
			"fieldtype": "Select",
			"options": "None\nbar\nline\npie\ndonut",
			"default": "line",
			hide_in_filter: 1,
			"on_change": function (query_report) { },
		},
		
		{
			"fieldname": "show_summary",
			"label": __("Show Summary"),
			"fieldtype": "Check",
			"default": 1,
			hide_in_filter: 1,
			"on_change": function (query_report) { },
		},
		{
			"fieldname": "hide_empty_record",
			"label": __("Hide Empty Record"),
			"fieldtype": "Check",
			"default": 0,
			"on_change": function (query_report) { },
			hide_in_filter: 1,
		},
 
	],
	onload: function (report) {
		report.page.add_inner_button("Preview Report", function () {
			frappe.query_report.refresh();
		});

		setLinkField();




	},
	"formatter": function (value, row, column, data, default_formatter) {
		const origninal_value = value || 0
		value = default_formatter(value, row, column, data);



		value = value.toString().replace("style='text-align: right'", "style='text-align: " + column.align + "'");

 
			
		if (origninal_value == 0 || origninal_value == "0") {
			return "<div style='text-align:" + (column.align || "left") + ";'>-</div>"
		}
	 




		if ((data && data.is_group == 1) || (data && data.is_total_row == 1) || (data && data.is_occupancy_row == 1)) {

			value = $(`<span>${value}</span>`);

			var $value = $(value).css("font-weight", "bold");


			value = $value.wrap("<p></p>").parent().html();
		}


		return value;
	},

};



function setLinkField() {
	const property = frappe.query_report.get_filter_value("property")
	if (property) {
		 
		
		// frappe.query_report.get_filter('room_type').df.get_query = function () {
		// 	return {
		// 		filters: {
		// 			"property": property
		// 		}
		// 	};
		// };

		 
	}

}